File: apps/backend/src/lib/corridors.ts
/**
 * Payout corridors: everything that differs between the countries VessPay pays
 * into, in one place, so a rail-specific rule never has to be rediscovered by
 * grepping for 'GHS'.
 *
 * Verified against the WeWire stage sandbox on 2026-09-07:
 *   - GET /v1/banks?currency=GHS returns 28 institutions, MOMO and BANK.
 *   - GET /v1/banks?currency=NGN returns 422 institutions, *every one* BANK.
 * Nigeria therefore has no mobile money channel: the wallets a Nigerian would
 * call mobile money (OPay, PalmPay, Moniepoint, Kuda, Paga) are BANK entries
 * addressed by a 10-digit NUBAN, not by a phone number.
 */

import type { PayoutChannel } from './wewire-institutions';

export interface Corridor {
  /** ISO 3166-1 alpha-2, as the app and the travel profile use it. */
  country: string;
  /** ISO 3166-1 alpha-3, as WeWire requires it on beneficiaries. */
  alpha3: string;
  name: string;
  /** Payout currency. */
  currency: string;
  /** Symbol for user-facing amounts. */
  symbol: string;
  dialCode: string;
  /** Channels this corridor can actually be paid over. */
  channels: PayoutChannel[];
  /** Local bank account number length, in digits. */
  accountNumber: { min: number; max: number };
  /**
   * National mobile money MSISDN pattern, or null for a corridor with no
   * mobile money channel.
   */
  msisdnPattern: RegExp | null;
  /**
   * WeWire's flat processor fee per disbursement, denominated in the payout
   * asset. Overridable per currency with WEWIRE_PROCESSOR_FEE_<CURRENCY>.
   */
  processorFee: number;
  /** True once a payout has actually been observed settling on this rail. */
  feeConfirmed: boolean;
}

export const CORRIDORS: Corridor[] = [
  {
    country: 'GH',
    alpha3: 'GHA',
    name: 'Ghana',
    currency: 'GHS',
    symbol: 'GH₵',
    dialCode: '+233',
    channels: ['MOBILE_MONEY', 'BANK'],
    // Ghana account numbers run 8-20 digits depending on the bank, so the
    // check stays deliberately loose.
    accountNumber: { min: 8, max: 20 },
    msisdnPattern: /^0[235]\d{8}$/,
    // Observed on every sandbox payout: a flat 5, denominated in the payout
    // asset (GHST), regardless of the amount sent.
    processorFee: 5.0,
    feeConfirmed: true,
  },
  {
    country: 'NG',
    alpha3: 'NGA',
    name: 'Nigeria',
    currency: 'NGN',
    symbol: '₦',
    dialCode: '+234',
    // No MOMO institutions exist for NGN -- see the file header.
    channels: ['BANK'],
    // NUBAN is exactly 10 digits.
    accountNumber: { min: 10, max: 10 },
    msisdnPattern: null,
    // PLACEHOLDER, not yet measured: no NGN payout has settled yet, so this is
    // an estimate deliberately set high enough to over-collect rather than
    // under-collect (the quote is reconciled against WeWire's actual fee once
    // the payout is sent). Replace with the observed value -- and set
    // feeConfirmed -- after the first live NGN disbursement.
    processorFee: 400.0,
    feeConfirmed: false,
  },
];

const DEFAULT_CORRIDOR = CORRIDORS[0];

/** Every alias the app has historically used to name a corridor. */
const CORRIDOR_LOOKUP: Record<string, Corridor> = {};
for (const corridor of CORRIDORS) {
  for (const alias of [
    corridor.country,
    corridor.alpha3,
    corridor.currency,
    corridor.name.toUpperCase(),
  ]) {
    CORRIDOR_LOOKUP[alias] = corridor;
  }
}

/**
 * Resolves a country code, alpha-3, currency or country name to its corridor.
 * Returns null rather than guessing, so a caller can decide whether an unknown
 * destination is an error or a reason to fall back to Ghana.
 */
export function findCorridor(input?: string | null): Corridor | null {
  if (!input || typeof input !== 'string') return null;
  return CORRIDOR_LOOKUP[input.trim().toUpperCase()] || null;
}

/**
 * Resolves a corridor, falling back to Ghana. Use where the pre-corridor code
 * already defaulted to Ghana so behaviour is unchanged for existing callers.
 */
export function getCorridor(input?: string | null): Corridor {
  return findCorridor(input) || DEFAULT_CORRIDOR;
}

/** The payout currency for a country, e.g. 'GH' -> 'GHS'. */
export function currencyForCountry(country?: string | null): string {
  return getCorridor(country).currency;
}

export function isSupportedCorridor(input?: string | null): boolean {
  return findCorridor(input) !== null;
}

/** Comma separated corridor currencies, for validation error messages. */
export function supportedPayoutCurrencies(): string {
  return CORRIDORS.map((c) => c.currency).join(', ');
}

/**
 * WeWire's flat processor fee for a corridor, in the payout currency.
 * WEWIRE_PROCESSOR_FEE_<CURRENCY> overrides it, which is how the Nigerian
 * placeholder gets corrected without a deploy.
 */
export function processorFeeFor(currency: string): number {
  const corridor = findCorridor(currency);
  const code = (corridor?.currency || currency).trim().toUpperCase();

  const envFee = process.env[`WEWIRE_PROCESSOR_FEE_${code}`];
  if (envFee && !isNaN(parseFloat(envFee))) return parseFloat(envFee);

  return corridor?.processorFee ?? 0;
}

/** Whether a corridor can be paid over a given channel. */
export function supportsChannel(input: string | null | undefined, channel: PayoutChannel): boolean {
  const corridor = findCorridor(input);
  return corridor ? corridor.channels.includes(channel) : false;
}

File: apps/mobile/lib/core/config/corridors.dart
/// Payout corridors: the client-side mirror of the backend's lib/corridors.ts.
///
/// Everything that differs between the countries VessPay pays into lives here,
/// so a rail-specific rule is never spelled out inline in a screen. The server
/// remains the authority -- GET /api/travel/destinations carries `symbol`,
/// `channels` and `payoutAvailable` -- and this table is what the app falls
/// back to before that call returns, or when it fails.
class Corridor {
  final String country;
  final String name;
  final String currency;

  /// Symbol for user-facing amounts, e.g. 'GH₵'.
  final String symbol;

  /// International dialling prefix, e.g. '+233'.
  final String dialCode;

  /// Channels this corridor can be paid over: 'MOBILE_MONEY' and/or 'BANK'.
  final List<String> channels;

  /// Local bank account number length, in digits.
  final int accountMinDigits;
  final int accountMaxDigits;

  /// National mobile money number pattern, or null where the corridor has no
  /// mobile money channel.
  final RegExp? msisdnPattern;

  /// Whether a payout can actually be sent today. Nigeria's reference data all
  /// works -- banks, account lookup, beneficiary registration -- but the
  /// provider exposes no NGN payout rail, so the destination is offered as
  /// coming soon rather than letting a user reach a dead Send button.
  final bool payoutAvailable;

  const Corridor({
    required this.country,
    required this.name,
    required this.currency,
    required this.symbol,
    required this.dialCode,
    required this.channels,
    required this.accountMinDigits,
    required this.accountMaxDigits,
    required this.msisdnPattern,
    required this.payoutAvailable,
  });

  bool get hasMobileMoney => channels.contains('MOBILE_MONEY');
  bool get hasBank => channels.contains('BANK');

  /// True when the corridor offers exactly one way to be paid, in which case
  /// the payment-type selector is noise and is hidden.
  bool get isSingleChannel => channels.length == 1;

  String get flagEmoji {
    switch (country.toUpperCase()) {
      case 'GH':
        return '🇬🇭';
      case 'NG':
        return '🇳🇬';
      case 'KE':
        return '🇰🇪';
      default:
        return '🌍';
    }
  }

  /// Default institution code for a channel: the first thing a user is most
  /// likely to want, and always a code the payout rail actually recognises.
  ///
  /// Empty where the provider publishes no institution list for the corridor
  /// (Kenya) -- an invented code would be worse than none, because it would
  /// look selectable and be rejected on submission.
  String defaultNetworkFor(String channel) {
    switch (country) {
      case 'GH':
        return channel == 'MOBILE_MONEY' ? 'MTN' : 'GCB';
      case 'NG':
        return '100004'; // OPay; NGN has no mobile money channel
      default:
        return '';
    }
  }

  /// Human-readable account number rule, for validation messages.
  String get accountRuleText => accountMinDigits == accountMaxDigits
      ? '$accountMinDigits digits'
      : '$accountMinDigits-$accountMaxDigits digits';
}

/// Ghana's mobile money number pattern.
final RegExp kGhanaMsisdnPattern = RegExp(r'^0[235]\d{8}$');

final Corridor ghanaCorridor = Corridor(
  country: 'GH',
  name: 'Ghana',
  currency: 'GHS',
  symbol: 'GH₵',
  dialCode: '+233',
  channels: const ['MOBILE_MONEY', 'BANK'],
  // Ghana account numbers run 8-20 digits depending on the bank.
  accountMinDigits: 8,
  accountMaxDigits: 20,
  msisdnPattern: kGhanaMsisdnPattern,
  payoutAvailable: true,
);

final Corridor nigeriaCorridor = Corridor(
  country: 'NG',
  name: 'Nigeria',
  currency: 'NGN',
  symbol: '₦',
  dialCode: '+234',
  // No mobile money institutions exist for NGN: OPay, PalmPay, Moniepoint and
  // Kuda are all reached as banks, by 10-digit NUBAN.
  channels: const ['BANK'],
  accountMinDigits: 10,
  accountMaxDigits: 10,
  msisdnPattern: null,
  // The provider has no NGN payout rail yet -- see lib/corridors.ts.
  payoutAvailable: false,
);

/// Safaricom and Airtel MSISDNs: 07xxxxxxxx and 01xxxxxxxx.
final RegExp kKenyaMsisdnPattern = RegExp(r'^0[17]\d{8}$');

final Corridor kenyaCorridor = Corridor(
  country: 'KE',
  name: 'Kenya',
  currency: 'KES',
  symbol: 'KSh',
  dialCode: '+254',
  // Not from the provider: WeWire publishes no Kenyan institution list, so
  // this is the market shape (M-Pesa first, banks second) rather than
  // anything verified.
  channels: const ['MOBILE_MONEY', 'BANK'],
  accountMinDigits: 6,
  accountMaxDigits: 20,
  msisdnPattern: kKenyaMsisdnPattern,
  // Rates resolve (USD/KES direct, GBP and EUR through the USD cross) but
  // there is no institution list, no account lookup and no payout rail.
  payoutAvailable: false,
);

final List<Corridor> kCorridors = [
  ghanaCorridor,
  nigeriaCorridor,
  kenyaCorridor,
];

/// Resolves a country code, currency or country name to its corridor,
/// defaulting to Ghana the way the app did before a second corridor existed.
Corridor corridorFor(String? input) {
  if (input == null || input.trim().isEmpty) return ghanaCorridor;
  final key = input.trim().toUpperCase();
  for (final c in kCorridors) {
    if (c.country == key || c.currency == key || c.name.toUpperCase() == key) {
      return c;
    }
  }
  return ghanaCorridor;
}

/// Returns null rather than defaulting, for callers that need to know whether
/// a destination is served at all.
Corridor? findCorridor(String? input) {
  if (input == null || input.trim().isEmpty) return null;
  final key = input.trim().toUpperCase();
  for (final c in kCorridors) {
    if (c.country == key || c.currency == key || c.name.toUpperCase() == key) {
      return c;
    }
  }
  return null;
}

/// Formats an amount in a corridor's currency, e.g. 'GH₵150.00'.
String formatCorridorAmount(String currencyOrCountry, num amount) {
  final corridor = corridorFor(currencyOrCountry);
  return '${corridor.symbol}${amount.toStringAsFixed(2)}';
}
